8262017: C2: assert(n != __null) failed: Bad immediate dominator info.

chhagedorn opened the pull request:
The wrong dominance information is caused by a `ConvL2I` node whose input gets a wrong type which is out of the range of the `ConvL2I`. As a consequence, the node becomes top and some data nodes are folded away which eventually results in the dominance assertion failure.

The wrong type information can be traced back to the overflow/underflow handling in range check elimination. Originally, the code in `PhaseIdealLoop::add_constraint()` covered the following special case for the adjustment of the pre loop limit:
https://github.com/openjdk/jdk/blob/cd33abb136d415455090cef4fe6211d9e6940948/src/hotspot/share/opto/loopTransform.cpp#L2423-L2456

This code, however, was removed later by accident:
https://github.com/openjdk/jdk/commit/afd852ccb8f8d36c721963ba68de695ce3e0e23d#diff-6a59f91cb710d682247df87c75faf602f0ff9f87e2855ead1b80719704fbedffL2370-L2394

This was only revealed after enabling more precise type information for phi nodes in [JDK-8257813](https://bugs.openjdk.java.net/browse/JDK-8257813) after which the testcase starts to fail.

The proposed fix is straight forward to reintroduce the wrongly removed code and adapt it to long instead of int.

Thanks,
Christian

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262017](https://bugs.openjdk.java.net/browse/JDK-8262017): C2: assert(n != __null) failed: Bad immediate dominator info.


### Reviewers
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4362/head:pull/4362` \
`$ git checkout pull/4362`

Update a local copy of the PR: \
`$ git checkout pull/4362` \
`$ git pull https://git.openjdk.java.net/jdk pull/4362/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4362`

View PR using the GUI difftool: \
`$ git pr show -t 4362`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4362.diff">https://git.openjdk.java.net/jdk/pull/4362.diff</a>

</details>

rwestrel commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2425,6 +2425,25 @@ void PhaseIdealLoop::add_constraint(jlong stride_con, jlong scale_con, Node* off
     //     else /* scale > 0 and stride < 0 */
     //       I > (low_limit-(offset+1))/scale
     //   )
+    if (low_limit->get_long() == -max_jint) {
```
There was similar code in both branches of the if ((stride_con^scale_con) >= 0) { but you're adding it back in only one of the branches. Is it not needed in the other one?

chhagedorn replied:
I think I just missed on that one. In theory, we would just do more iterations in the pre-loop instead of the main-loop when underflowing and it will be fine. But I'm not sure if we eventually run into similar problems as with the other branch (I could not trigger that though). So, I reintroduced the code for this branch as well.

rwestrel commented:
Roughly, the problem as I understand it from your description/patch is that: new bounds are computed using long arithmetic which causes bounds to not fit in an int and to be incorrect once the bounds are casted to an int. Is that it? Can you give some more details?

So the previous patch dropped some overflow checks under the assumption, I suppose, that there was no need for them because of the move to long arithmetic and now we're finding that it's doesn't really work that way. Was long arithmetic the right move then? With the previous code added back, does it still solve problems? Or was that approach fundamentally flawed?

chhagedorn commented:
I had to dig a little bit deeper to fully get that code. I was missing some points before but now my improved understanding is the following:
The approach with longs only partially solved the underflow/overflow problem addressed in `PhaseIdealLoop::adjust_limit`. The goal was to clamp the limits like that:
https://github.com/openjdk/jdk/blob/94d0b0f9810bd1a8da06ec267a1c7589d6cb756b/src/hotspot/share/opto/loopTransform.cpp#L2334-L2338
We achieve that by doing a long version of the inner min/max operation with cmoves. For example, let's look at the first inner MAX: `MAX(limit, min_jint)`: When `limit` is less than `min_jint` (integer underflow), then we will use `min_jint` and avoid the underflow problem. However, this only fixes the underflow case but not the overflow case of `limit` (which is the root cause of this bug). In this bug, `limit` is greater than `max_jint`. Therefore, `MAX(limit, min_jint)` takes `limit` (long comparison) which is then converted back to int resulting in a very small negative number near `min_jint` due to integer overflow. This value is less than `old_limit`, so the outer integer based MIN operation takes that value resulting in a wrong limit. This causes some folding operations later and we end up with an assertion failure. The other case in the comment above is similar but just reverted: integer overflow is handled but underflow is not.

Based on that I found that the previous fix was only partially working and has not fixed the root cause. I reverted it and pushed a new fix for it addressing the missing overflow/underflow handling described above. I now switched to a long version of min/max for the outer operation (see comment above) with cmoves as well and only then convert back to int. AFAICT this fixes the wrong limit computations and handles both integer overflow and underflow correctly. I tried to improve the comments in `PhaseIdealLoop::adjust_limit` to make the intentions of the code more clear.

rwestrel commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2331,23 +2331,41 @@ Node* PhaseIdealLoop::adjust_limit(bool is_positive_stride, Node* scale, Node* o
+  register_new_node(inner_result_int, pre_ctrl);
+  limit = new CMoveINode(bol, old_limit, inner_result_int, TypeInt::INT);
   register_new_node(limit, pre_ctrl);
+  C->print_method(PHASE_ADD_UNSAFE_BARRIER, 3);
```
Does that belong here?

chhagedorn replied:
Removed, was a leftover from debugging

rwestrel commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2331,23 +2331,41 @@ Node* PhaseIdealLoop::adjust_limit(bool is_positive_stride, Node* scale, Node* o
-
-  limit = is_positive_stride ? (Node*)(new MinINode(old_limit, limit))
-                             : (Node*)(new MaxINode(old_limit, limit));
+  Node* inner_result_long = new CMoveLNode(bol, limit, _igvn.longcon(is_positive_stride ? min_jint : max_jint), TypeLong::LONG);
```
Could any of that code take advantage of MaxNode::signed_max()/signed_min?

chhagedorn replied:
I think we can replace the first one with it. But as I think we need to keep the second one as it is as it directly uses `CMovI`.

rwestrel commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2331,22 +2331,39 @@ Node* PhaseIdealLoop::adjust_limit(bool is_positive_stride, Node* scale, Node* o
+  // We use helper methods for inner MINL/MAXL which return CMoveL nodes to keep a long value for the outer MINL/MAXL comparison:
+  Node* inner_result_long;
+  if (is_positive_stride) {
+    inner_result_long = MaxNode::signed_max(limit, _igvn.longcon(min_jint), TypeLong::LONG, _igvn);
```
You still need to register the new nodes with PhaseIdealLoop (what you achieved before with register_new_node()). set_subtree_ctrl() is what you need I think.

rwestrel reviewed (approved):
That looks good to me.

vnkozlov reviewed (approved):
Looks good.

chhagedorn commented:
Thanks @rwestrel and @vnkozlov for your reviews! I'm closing this PR and will reopen a new one for JDK 17.
